SICP_book/chapter_2_1.metta

```diff
@@ -0,0 +1,793 @@
+    (println! ((numer $x)/(denom $x))))
+
+(= (one-half) (make-rat 1 2))
+
+(= (numer $x) (let $x ($a . $b) $a))
+
+(= (denom $x)
+    (let $x ($a . $b) $b))
+
+; This line should print [(1 / 2)]
+;!(print-rat (make-rat 1 2))
+
+(= (one-third) (make-rat 1 3))
+
+;!(print-rat (add-rat (one-half) (one-third)))
+;!(print-rat (mul-rat (one-half) (one-third)))
+;!(print-rat (add-rat (one-third) (one-third)))
+
+(= (remainder $x $y) (% $x $y))
+
+(= (gcd $a $b)
+    (if (== $b 0)
+        $a
+        (gcd $b (remainder $a $b))))
+
+(= (make-rat2 $n $d)
+    (let $g (gcd $n $d)
+        (make-pair (/ $n $g) (/ $d $g))))
+
+(= (add-rat2 $x $y)
+  (make-rat2 (+ (* (numer $x) (denom $y))
+               (* (numer $y) (denom $x)))
+            (* (denom $x) (denom $y))))
+
+; In Scheme it (2 / 3) will be printed, but in our case - (2.0 / 3.0)
+;!((print-rat (add-rat2 (one-third) (one-third))))
+
+; Exercise 2.1.
+;
+; Define a better version of make-rat that handles both positive and negative arguments.
+; make-rat should normalize the sign so that if the rational number is positive, both the numerator and
+; denominator are positive, and if the rational number is negative, only the numerator is negative.
+
+(= (Abs $x) (if (< $x 0) (* $x -1) $x))
+
+(= (better_make-rat $n $d)
+    (let $g (gcd $n $d)
+        (if (or (< $n 0) (< $d 0))
+            (make-pair (/ $n $g) (Abs (/ $d $g)))
+            (make-pair (Abs (/ $n $g)) (Abs (/ $d $g))))))
+
+;!(print-rat (better_make-rat 1 2)) ; 1.0 / 2.0
+;!(print-rat (better_make-rat -1 2)) ; - 1.0 / 2.0
+;!(print-rat (better_make-rat 1 -2)) ; - 1.0 / 2.0
+;!(print-rat (better_make-rat -1 -2)); 1.0 / 2.0
+
+; -----------------------End of Exercise 2.1----------------------------
+
+; Exercise 2.2.
+;
+; Consider the problem of representing line segments in a plane. Each segment is
+; represented as a pair of points: a starting point and an ending point. Define a constructor
+; make-segment and selectors start-segment and end-segment that define the representation of
+; segments in terms of points. Furthermore, a point can be represented as a pair of numbers: the x
+; coordinate and the y coordinate. Accordingly, specify a constructor make-point and selectors
+; x-point and y-point that define this representation. Finally, using your selectors and constructors,
+; define a procedure midpoint-segment that takes a line segment as argument and returns its midpoint
+; (the point whose coordinates are the average of the coordinates of the endpoints). To try your procedures,
+; you'll need a way to print points
+
+(= (make-segment $start $end) (cons-atom $start $end))
+
+(= (start-segment $seg) (car-atom $seg))
+(= (end-segment $seg) (cdr-atom $seg))
+
+(= (make-point $x $y) (make-pair $x $y))
+
+(= (x-point $pt) (let ($x . $y) $pt $x))
+(= (y-point $pt) (let ($x . $y) $pt $y))
+
+!(assertEqual
+    (make-point 5 2)
+    (5 . 2))
+
+!(assertEqual
+    (make-segment (make-point 1 2) (make-point 5 7))
+    ((1 . 2) 5 . 7))
```

Comment:
I believe it is because `cons` in Scheme and `cons-atom` in MeTTa has different semantics. In Scheme `cons` makes a pair from two values. In MeTTa `cons-atom` builds a list from head value and tail list. Thus using `cons-atom` instead of `cons` is not valid. Same difference is present in `car-atom` and `cdr-atom`.

Reply:
Well, I've used scheme-like result of cons indeed. I'll replace it with something better.

Reply:
Ok, I'll avoid using cons-atom and car-/cdr-atom here. At least here it is easily avoidable.